### ISSUE TITLE
"internal compiler error" (invalid instruction) when bootstrapping Go toolchain (gc) using gccgo

### ISSUE THREAD

**Issue by bagasme** [golang/go#49312]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16.5 gccgo (GCC) 11.2.0 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes (I'm using latest gccgo (GCC) release).

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/bagas/.cache/go-build"
GOENV="/home/bagas/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/bagas/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/bagas/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/opt/llvm"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/opt/llvm/libexec/gcc/x86_64-pc-linux-gnu/11.2.0"
GOVCS=""
GOVERSION="go1.16.5 gccgo (GCC) 11.2.0"
GCCGO="/opt/llvm/bin/gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1807043512=/tmp/go-build -gno-record-gcc-switches -funwind-tables"
</pre></details>

### What did you do?

I'm experimenting on bootstrapping Go 1.17.2 toolchain using gccgo (by following [the instructions](https://golang.org/doc/install/source#bootstrapFromGccgo)). I use GCC installed in `/opt/llvm` (which also used to bootstrap LLVM, hence the prefix).

### What did you expect to see?

Bootstrapping success.

### What did you see instead?

Bootstrapping failed with "internal compiler error" during `Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1` stage. The log for that stage says:

```
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
runtime
runtime/internal/math
internal/abi
runtime/internal/atomic
internal/bytealg
internal/goexperiment
runtime/internal/sys
internal/cpu
/tmp/golang/src/runtime/internal/sys/intrinsics_common.go:142:2: "".Ctz64: invalid instruction: 00049 (/tmp/golang/src/runtime/internal/sys/intrinsics.go:44)	MOVQ	CX, "".~r1+16(DX)
/tmp/golang/src/runtime/internal/sys/intrinsics_common.go:142:2: "".Ctz32: invalid instruction: 00042 (/tmp/golang/src/runtime/internal/sys/intrinsics.go:54)	MOVQ	AX, "".~r1+16(DX)
/tmp/golang/src/runtime/internal/sys/intrinsics_common.go:142:2: "".Len64: invalid instruction: 00105 (/tmp/golang/src/runtime/internal/sys/intrinsics_common.go:61)	MOVQ	DX, "".n+40(CX)
/tmp/golang/src/runtime/internal/sys/intrinsics_common.go:142:2: "".LeadingZeros64: invalid instruction: 00027 (/tmp/golang/src/runtime/internal/sys/intrinsics_common.go:130)	MOVQ	CX, "".~r1+16(AX)

go tool dist: FAILED: /tmp/golang/pkg/tool/linux_amd64/compile -std -pack -o /tmp/go-tool-dist-093252194/runtime/internal/sys/_go_.a -p runtime/internal/sys -importcfg /tmp/go-tool-dist-093252194/runtime/internal/sys/importcfg /tmp/golang/src/runtime/internal/sys/arch.go /tmp/golang/src/runtime/internal/sys/arch_amd64.go /tmp/golang/src/runtime/internal/sys/intrinsics.go /tmp/golang/src/runtime/internal/sys/intrinsics_common.go /tmp/golang/src/runtime/internal/sys/sys.go /tmp/golang/src/runtime/internal/sys/zgoarch_amd64.go /tmp/golang/src/runtime/internal/sys/zgoos_linux.go /tmp/golang/src/runtime/internal/sys/zversion.go: exit status 2
go tool dist: open /tmp/go-tool-dist-093252194/runtime/internal/sys/_go_.a: no such file or directory
go tool dist: open /tmp/go-tool-dist-093252194/internal/cpu/_go_.a: no such file or directory
/tmp/golang/src/internal/cpu/cpu_x86.go:46:17: internal compiler error: 'doinit': panic during regalloc while compiling doinit:

interface conversion: ssa.Location is nil, not *	bootstrap_1cmd_1compile_1internal_1ssa	ssa.Register

goroutine 1 [running]:
bootstrap/cmd/compile/internal/ssa.Compile..func1
	/tmp/golang/src/cmd/compile/internal/ssa/compile.go:44
panic
	:0
bootstrap_1cmd_1compile_1internal_1ssa.regAllocState.regalloc
	/tmp/golang/src/cmd/compile/internal/ssa/regalloc.go:1578
bootstrap/cmd/compile/internal/ssa.regalloc
	/tmp/golang/src/cmd/compile/internal/ssa/regalloc.go:148
bootstrap_1cmd_1compile_1internal_1ssa.Compile
	/tmp/golang/src/cmd/compile/internal/ssa/compile.go:96
bootstrap/cmd/compile/internal/ssagen.buildssa
	/tmp/golang/src/cmd/compile/internal/ssagen/ssa.go:642
bootstrap_1cmd_1compile_1internal_1ssagen.Compile
	/tmp/golang/src/cmd/compile/internal/ssagen/pgen.go:165
bootstrap/cmd/compile/internal/gc.compileFunctions..func6
	/tmp/golang/src/cmd/compile/internal/gc/compile.go:153
bootstrap/cmd/compile/internal/gc.compileFunctions..func2
	/tmp/golang/src/cmd/compile/internal/gc/compile.go:125
bootstrap/cmd/compile/internal/gc.compileFunctions..func5
	/tmp/golang/src/cmd/compile/internal/gc/compile.go:152
bootstrap/cmd/compile/internal/gc.compileFunctions
	/tmp/golang/src/cmd/compile/internal/gc/compile.go:163
bootstrap_1cmd_1compile_1internal_1gc.Main
	/tmp/golang/src/cmd/compile/internal/gc/main.go:287
main.main
	/tmp/golang/src/cmd/compile/main.go:55



Please file a bug report including a short program that triggers the error.
https://golang.org/issue/new

go tool dist: FAILED: /tmp/golang/pkg/tool/linux_amd64/compile -std -pack -o /tmp/go-tool-dist-093252194/internal/cpu/_go_.a -p internal/cpu -importcfg /tmp/go-tool-dist-093252194/internal/cpu/importcfg -asmhdr /tmp/go-tool-dist-093252194/internal/cpu/go_asm.h -symabis /tmp/go-tool-dist-093252194/internal/cpu/symabis /tmp/golang/src/internal/cpu/cpu.go /tmp/golang/src/internal/cpu/cpu_amd64.go /tmp/golang/src/internal/cpu/cpu_x86.go: exit status 2
```

The issue doesn't appear when I bootstrap using Go 1.16.5 (gc toolchain, same Go version as gccgo).

**Comment by thanm** [comment]
Thanks for raising this issue. 

I am having trouble reproducing this problem. I tried first with tip gccgo, no issues there. I also built a fresh copy of gccgo using the GCC V11 release branch, and that seems to work for bootstrapping 1.17.2 for me.

Could you please post the exact command that you used, and if you could please run the "make.bash" with additional flags "-d -d" that would be great. E.g. 

GOROOT_BOOTSTRAP=/your/gccgo/install/here  bash make.bash -d -d 1> err.txt 2>&1

**Comment by ianlancetaylor** [comment]
Probably a dup of #47771.  If so, it will be fixed in the GCC 11.3 release.

**Comment by bagasme** [comment]
@ianlancetaylor This isn't duplicate of aforementioned issue. I have `invalid instruction` error instead of segfault/panic.

**Comment by bagasme** [comment]
@thanm When bootstrapping I use:

```
LD_LIBRARY_PATH=/opt/llvm/lib64 GOROOT_BOOTSTRAP=/opt/llvm ./make.bash -v
```

where GCC version I used is 11.2.0 (gccgo 1.16.5) installed at `/opt/llvm` (I intended to build GCC for bootstrapping LLVM, but beyond the scope of this issue). The host is Ubuntu 20.04.

**Comment by thanm** [comment]
I don't have access to a Ubuntu 20.04 system, so I'm not sure whether this is going to help me much in reproducing. Can you share the exact commands used to install GCC? I am wondering if there is something special about the version built for the distro, or something like that.

**Comment by thanm** [comment]
OK, I downloaded the gcc 11.2 sources, built gccgo from those, and was able to reproduce the problem.

It is indeed as Ian mentioned: 11.2 is missing the fix for issue #47771 -- I verified this by visually inspecting the source (expressions.cc). Please wait for 11.3 to be released, or alternatively you can build gccgo from source. Thanks.

**Comment by ianlancetaylor** [comment]
> I have invalid instruction error instead of segfault/panic.

For the record, the problem was that gccgo had a bug that caused it to miscompile the gc compiler such that the gc compiler generated invalid code.  Essentially any kind of failure was possible.

Sorry for the trouble.
